utils/frame/rpc-utils/src/lib.rs

```diff
@@ -0,0 +1,245 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020-2022 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Shared JSON-RPC client related code and abstractions.
+
+use async_trait::async_trait;
+use serde::de::DeserializeOwned;
+use sp_runtime::traits::{Block as BlockT, Header as HeaderT};
+use std::collections::VecDeque;
+
+pub use jsonrpsee::{
+	core::client::{ClientT, Subscription, SubscriptionClientT},
+	rpc_params,
+	ws_client::{WsClient, WsClientBuilder},
+};
+pub use sc_rpc_api::{
+	author::AuthorApiClient as AuthorApi, chain::ChainApiClient as ChainApi,
+	child_state::ChildStateApiClient as ChildStateApi, dev::DevApiClient as DevApi,
+	offchain::OffchainApiClient as OffchainApi, state::StateApiClient as StateApi,
+	system::SystemApiClient as SystemApi,
+};
+
+/// Create a new `WebSocket` connection with shared settings.
+pub async fn ws_client(uri: impl AsRef<str>) -> Result<WsClient, String> {
+	WsClientBuilder::default()
+		.max_request_body_size(u32::MAX)
+		.request_timeout(std::time::Duration::from_secs(60 * 10))
+		.connection_timeout(std::time::Duration::from_secs(60))
+		.max_notifs_per_subscription(1024)
+		.build(uri)
+		.await
+		.map_err(|e| format!("`WsClientBuilder` failed to build: {:?}", e))
+}
+
+/// Abstraction over RPC calling for headers.
+#[async_trait]
+pub trait HeaderProvider<Block: BlockT>
+where
+	Block::Header: HeaderT,
+{
+	/// Awaits for the header of the block with hash `hash`.
+	async fn get_header(&self, hash: Block::Hash) -> Block::Header;
+}
+
+#[async_trait]
+impl<Block: BlockT> HeaderProvider<Block> for WsClient
+where
+	Block::Header: DeserializeOwned,
+{
+	async fn get_header(&self, hash: Block::Hash) -> Block::Header {
+		ChainApi::<(), Block::Hash, Block::Header, ()>::header(self, Some(hash))
+			.await
+			.unwrap()
+			.unwrap()
+	}
+}
+
+/// Abstraction over RPC subscription for finalized headers.
+#[async_trait]
+pub trait HeaderSubscription<Block: BlockT>
+where
+	Block::Header: HeaderT,
+{
+	/// Await for the next finalized header from the subscription.
+	///
+	/// Returns `None` if either the subscription has been closed or there was an error when reading
+	/// an object from the client.
+	async fn next_header(&mut self) -> Option<Block::Header>;
+}
+
+#[async_trait]
+impl<Block: BlockT> HeaderSubscription<Block> for Subscription<Block::Header>
```

Comment:
I think it makes most sense to have this one in substrate and not in jsonrpsee but maybe we could have a shared crate where because we need it subxt and perhaps somewhere else :P

We could probably implement it for `T` instead of `Block::Header` to work